In [1166]:
# DSC 540
# Weeks 5 & 6
# Summer 2024
# Weeks 5 & 6 Exercises
# Created By Matthew Heinen
# 7/6/24

# Chapter 7: 

In [1168]:
# Read in data file. 

# Import packages 
import pandas as pd

In [1169]:
# read in candyhieracrhy2017.csv. 
candy = pd.read_excel("candyhierarchy2017.xlsx")

/opt/anaconda3/lib/python3.11/site-packages/openpyxl/worksheet/_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


In [1170]:
# Display candy to verify read in 
candy.head()

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0     90258773            NaN        NaN     NaN         NaN   
1     90272821             No       Male      44        USA    
2     90272829            NaN       Male      49         USA   
3     90272840             No       Male      40          us   
4     90272841             No       Male      23         usa   

  Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                              NaN                NaN   
1                               NM                MEH   
2                         Virginia                NaN   
3                               or                MEH   
4                         exton pa                JOY   

  Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                       
1                                            DESPAIR                                       
2                                                NaN                                       
3                                            DESPAIR                                       
4                                            DESPAIR                                       

  Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
0                           NaN              NaN  ...               NaN   
1                           JOY              MEH  ...               NaN   
2                           NaN              NaN  ...               NaN   
3                           JOY              MEH  ...               NaN   
4                           JOY          DESPAIR  ...               NaN   

                                  Q9: OTHER COMMENTS      Q10: DRESS  \
0                                                NaN             NaN   
1  Bottom line is Twix is really the only candy w...  White and gold   
2                                                NaN             NaN   
3                             Raisins can go to hell  White and gold   
4                                                NaN  White and gold   

  Unnamed: 113 Q11: DAY Q12: MEDIA [Daily Dish] Q12: MEDIA [Science]  \
0          NaN      NaN                     NaN                  NaN   
1          NaN   Sunday                     NaN                  1.0   
2          NaN      NaN                     NaN                  NaN   
3          NaN   Sunday                     NaN                  1.0   
4          NaN   Friday                     NaN                  1.0   

  Q12: MEDIA [ESPN] Q12: MEDIA [Yahoo] Click Coordinates (x, y)  
0               NaN                NaN                      NaN  
1               NaN                NaN                 (84, 25)  
2               NaN                NaN                      NaN  
3               NaN                NaN                 (75, 23)  
4               NaN                NaN                 (70, 10)  

[5 rows x 120 columns]

In [1171]:
# print current size of df. 
candy.shape

(2460, 120)

## Filter Out Missing Data:

In [1173]:
# Use dropna to drop all columns that are completely empty. 
# Pass Axis 1 to drop columns. 

candy = candy.dropna(axis =1, how ='all')

# Print Candy shape 
candy.shape

(2460, 120)

Here we see that no columns were completely empty. 

In [1175]:
# Use dropna to drop all rows that are completely empty. 
# pass how = "all to drop rows that are all na

candy = candy.dropna(how='all')

# Print Candy shape 
candy.shape

(2460, 120)

Here we can see that there are no rows that are completly empty. 

In [1177]:
# Evaluate age 
# Test if a value in a column is numeric and remove the rows
# use errors='coerce' to leave null values as NaN
candy['Q3: AGE'] = pd.to_numeric(candy['Q3: AGE'], errors='coerce')  


# Remove rows with NaN values in column 'Q3: AGE'
# Use drop NA
candy = candy.dropna(subset=['Q3: AGE'])  

# print size 
print(candy.shape)


(2352, 120)


Here we have removed rows where the value input in age was non numeric. 

# Remove Duplicates

In [1180]:
# check to see if any rows are duplicates or repeats
# use drop_duplicates
candy = candy.drop_duplicates()

#print shape
candy.shape

(2352, 120)

Here we see that now rows are complete duplicates. 

In [1182]:
# Check to see if any of the internal ID is duplicated, meaning we have multiple records for the same person. 
candy = candy.drop_duplicates(['Internal ID'])

#print shape
candy.shape

(2352, 120)

 Here we see that we have no duplicate rows or entries. 

## Discretization and Binning

lets create some bins based on the age demographic. 


In [1185]:
# Use describe to find some summary stats on age. 

print(candy['Q3: AGE'].describe())

count    2352.000000
mean       42.617304
std        23.846531
min         1.000000
25%        34.000000
50%        41.000000
75%        50.000000
max      1000.000000
Name: Q3: AGE, dtype: float64


Here we can see that we have some outliers. ages 1 and 1000 are for sure outliers. lets remove those values by filtering out df. 

In [1187]:
# Use .loc to filter outliers from the age column of the dataset 
candy = candy.loc[(candy['Q3: AGE'] >= 10) & (candy['Q3: AGE'] <= 100)]

#Print shape
print(candy.shape)

(2337, 120)


In [1188]:
# use pd.cut to create age bins in df. 

# Create bins and labels 
bins = [10, 18, 30, 50, 100] 
labels = ['10-18', '19-30', '31-50', '51-100']

# create the bins and add column called 'age_bins' 
candy['age_bins'] = pd.cut(candy['Q3: AGE'], bins=bins, labels=labels)

In [1189]:
# Verify the binning worked. 
candy.head

<bound method NDFrame.head of       Internal ID Q1: GOING OUT? Q2: GENDER  Q3: AGE Q4: COUNTRY  \
1        90272821             No       Male     44.0        USA    
2        90272829            NaN       Male     49.0         USA   
3        90272840             No       Male     40.0          us   
4        90272841             No       Male     23.0         usa   
6        90272853             No       Male     53.0         usa   
...           ...            ...        ...      ...         ...   
2455     90314359             No       Male     24.0         USA   
2456     90314580             No     Female     33.0         USA   
2457     90314634             No     Female     26.0        USA    
2458     90314658             No       Male     58.0         Usa   
2459     90314802             No     Female     66.0         usa   

     Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
1                                  NM                MEH   
2                            Virg

Here we have sucessfully added the age_bins bin. 

# Chapter 8: 

## Create Hierarchical Index

create a hierarchical index where the age_bins is above age. 

In [1193]:
# select a subset of candy for improved readability. 

c_small = candy[['age_bins', 'Q3: AGE', 'Q9: OTHER COMMENTS']]

#Drop null 
c_small = c_small.dropna()

# sort by age

c_small = c_small.sort_values(by='Q3: AGE')

# view
c_small.shape


(370, 3)

In [1194]:
# Use set index to create hierarchical index. 

c_small.set_index(['age_bins', 'Q3: AGE'], inplace=True)

c_small.head

<bound method NDFrame.head of                                                  Q9: OTHER COMMENTS
age_bins Q3: AGE                                                   
10-18    11.0     Some of these catagories don't make any sense ...
         11.0     I think that getting chick tracts would be fun...
         12.0                                                   Hey
         12.0                                I like food and horses
         12.0     I feel like i'm gonna get diabetes this Hallow...
...                                                             ...
51-100   71.0     As a presenter, I give the best stuff to obvio...
         71.0     This was the best scientific survey I have eve...
         77.0     I qualify to take this survey because I was on...
         88.0                    Good and Plenty all day every day.
         90.0     Why no spare change?  More vegan options?  Wha...

[370 rows x 1 columns]>

Here we see that we have created a hierarchy with the age and age bins column. 

In [1196]:
# Use sort incex to sort by age_bins
c_small.sort_index(level='age_bins')

# display
c_small

Q9: OTHER COMMENTS
age_bins Q3: AGE                                                   
10-18    11.0     Some of these catagories don't make any sense ...
         11.0     I think that getting chick tracts would be fun...
         12.0                                                   Hey
         12.0                                I like food and horses
         12.0     I feel like i'm gonna get diabetes this Hallow...
...                                                             ...
51-100   71.0     As a presenter, I give the best stuff to obvio...
         71.0     This was the best scientific survey I have eve...
         77.0     I qualify to take this survey because I was on...
         88.0                    Good and Plenty all day every day.
         90.0     Why no spare change?  More vegan options?  Wha...

[370 rows x 1 columns]

This heigheracy allows us to view each demographics other comments. 

## Reshape


In [1199]:
# Use stack to reshape the columns in the data to rows this allows us to see all results by age and bin
c_stack = c_small.stack()


print(c_stack.head(50))

age_bins  Q3: AGE                    
10-18     11.0     Q9: OTHER COMMENTS    Some of these catagories don't make any sense ...
                   Q9: OTHER COMMENTS    I think that getting chick tracts would be fun...
          12.0     Q9: OTHER COMMENTS                                                  Hey
                   Q9: OTHER COMMENTS                               I like food and horses
                   Q9: OTHER COMMENTS    I feel like i'm gonna get diabetes this Hallow...
          13.0     Q9: OTHER COMMENTS    As I do not know what around half of those ite...
          14.0     Q9: OTHER COMMENTS         I can't tell if this survey is a joke or not
                   Q9: OTHER COMMENTS    Since it says “mars” and not specifically a ma...
          15.0     Q9: OTHER COMMENTS    If you give out full-sized anything (candy, fo...
                   Q9: OTHER COMMENTS    A lot of the things i despair I didn't know, t...
          16.0     Q9: OTHER COMMENTS              I

This reshaping or stacking of hierarchical levels allows us to view comments by both age demographic and individual age.

# Chapter 10

## Grouping with Dicts/Series

Use dictionary or series methods to group by age_bin. 

In [1203]:
# use groupby to find the mean of each age bin in the candy df. 

grouped = candy['Q3: AGE'].groupby(candy['age_bins'])

/var/folders/xv/q0156cqs67s40wbxq6_vfykr0000gn/T/ipykernel_77206/3035731113.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped = candy['Q3: AGE'].groupby(candy['age_bins'])


In [1204]:
# Use mean function to display the mean of each age bin. 

grouped.mean()

age_bins
10-18     14.509804
19-30     26.586331
31-50     40.393347
51-100    58.418233
Name: Q3: AGE, dtype: float64

## Grouping with Index Levels

Now we will perform a similar task with a different method we will group by index levels using the hierarchy established in DataFrame c_small. We will do this to find the count of unique responses to question 9 in each age. 

In [1231]:
grouped = c_small.groupby(level = 'age_bins')


/var/folders/xv/q0156cqs67s40wbxq6_vfykr0000gn/T/ipykernel_77206/3434568292.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped = c_small.groupby(level = 'age_bins')


In [1235]:
grouped.describe()


Q9: OTHER COMMENTS         \
                      count unique   
age_bins                             
10-18                    13     13   
19-30                    42     42   
31-50                   209    209   
51-100                  106    106   

                                                                  
                                                        top freq  
age_bins                                                          
10-18     Some of these catagories don't make any sense ...    1  
19-30                               You folks are awesome!     1  
31-50     War on Wonka! Fuck Roald Dahl and everything h...    1  
51-100    while the JOY OTHER items reflect the best hal...    1

Now lets us grouping by index to find some descriptive stats on the ages of each age bin. 

In [1243]:
# Create new smaller df for readability. 

c_small_2 = candy[['age_bins', 'Q3: AGE']]

#Drop null 
c_small_2 = c_small_2.dropna()

# sort by age

c_small_2 = c_small_2.sort_values(by='Q3: AGE')

In [1245]:
# Create hierarchy

c_small_2.set_index(['age_bins'], inplace=True)



In [1249]:
c_small_2.groupby(level = 'age_bins').describe()

/var/folders/xv/q0156cqs67s40wbxq6_vfykr0000gn/T/ipykernel_77206/4266034155.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  c_small_2.groupby(level = 'age_bins').describe()


Q3: AGE                                                    
           count       mean       std   min   25%   50%   75%    max
age_bins                                                            
10-18       51.0  14.509804  2.532766  11.0  12.0  15.0  17.0   18.0
19-30      278.0  26.586331  3.064545  19.0  25.0  27.0  29.0   30.0
31-50     1473.0  40.393347  5.538554  31.0  36.0  40.0  45.0   50.0
51-100     532.0  58.418233  6.965775  51.0  53.0  57.0  62.0  100.0

Above we can see that we have found some the descriptive stats of the ages in each age bin using the hierarchical indexing method. 

# Chapter 11 

In [1532]:
# Import Pandas 
import pandas as pd

In [1534]:
# Read in Met Museum data set. 
# Use read_txt
# Separator is a comma
met = pd.read_csv("MetObjects.txt", sep=',')

/var/folders/xv/q0156cqs67s40wbxq6_vfykr0000gn/T/ipykernel_77206/1546250766.py:4: DtypeWarning: Columns (5,7,10,11,12,13,14,34,35,36,37,38,39,40,41,42,43,44,45,46) have mixed types. Specify dtype option on import or set low_memory=False.
  met = pd.read_csv("MetObjects.txt", sep=',')


In [1535]:
# Preview dataset. 
met.head(5)


Object Number  Is Highlight  Is Timeline Work  Is Public Domain  Object ID  \
0    1979.486.1         False             False             False          1   
1    1980.264.5         False             False             False          2   
2      67.265.9         False             False             False          3   
3     67.265.10         False             False             False          4   
4     67.265.11         False             False             False          5   

  Gallery Number         Department AccessionYear Object Name  \
0            NaN  The American Wing        1979.0        Coin   
1            NaN  The American Wing        1980.0        Coin   
2            NaN  The American Wing        1967.0        Coin   
3            NaN  The American Wing        1967.0        Coin   
4            NaN  The American Wing        1967.0        Coin   

                          Title  ... River Classification  \
0  One-dollar Liberty Head Coin  ...   NaN            NaN   
1  Ten-dollar Liberty Head Coin  ...   NaN            NaN   
2    Two-and-a-Half Dollar Coin  ...   NaN            NaN   
3    Two-and-a-Half Dollar Coin  ...   NaN            NaN   
4    Two-and-a-Half Dollar Coin  ...   NaN            NaN   

  Rights and Reproduction                                     Link Resource  \
0                     NaN  http://www.metmuseum.org/art/collection/search/1   
1                     NaN  http://www.metmuseum.org/art/collection/search/2   
2                     NaN  http://www.metmuseum.org/art/collection/search/3   
3                     NaN  http://www.metmuseum.org/art/collection/search/4   
4                     NaN  http://www.metmuseum.org/art/collection/search/5   

  Object Wikidata URL Metadata Date                                Repository  \
0                 NaN           NaN  Metropolitan Museum of Art, New York, NY   
1                 NaN           NaN  Metropolitan Museum of Art, New York, NY   
2                 NaN           NaN  Metropolitan Museum of Art, New York, NY   
3                 NaN           NaN  Metropolitan Museum of Art, New York, NY   
4                 NaN           NaN  Metropolitan Museum of Art, New York, NY   

  Tags Tags AAT URL Tags Wikidata URL  
0  NaN          NaN               NaN  
1  NaN          NaN               NaN  
2  NaN          NaN               NaN  
3  NaN          NaN               NaN  
4  NaN          NaN               NaN  

[5 rows x 54 columns]

In [1536]:
# Show columns names. 

# Print column names
for column in met.columns:
    print("Column:", column)

# Print index
print("Index:", met.index)

Column: Object Number
Column: Is Highlight
Column: Is Timeline Work
Column: Is Public Domain
Column: Object ID
Column: Gallery Number
Column: Department
Column: AccessionYear
Column: Object Name
Column: Title
Column: Culture
Column: Period
Column: Dynasty
Column: Reign
Column: Portfolio
Column: Constituent ID
Column: Artist Role
Column: Artist Prefix
Column: Artist Display Name
Column: Artist Display Bio
Column: Artist Suffix
Column: Artist Alpha Sort
Column: Artist Nationality
Column: Artist Begin Date
Column: Artist End Date
Column: Artist Gender
Column: Artist ULAN URL
Column: Artist Wikidata URL
Column: Object Date
Column: Object Begin Date
Column: Object End Date
Column: Medium
Column: Dimensions
Column: Credit Line
Column: Geography Type
Column: City
Column: State
Column: County
Column: Country
Column: Region
Column: Subregion
Column: Locale
Column: Locus
Column: Excavation
Column: River
Column: Classification
Column: Rights and Reproduction
Column: Link Resource
Column: Object W

## Convert Between String and Date Time

In [1538]:
# View Date column values. 
# View Object Date, Begin Date and End Date
# View Artist Date, Begin Date and End Date

# create smaller met_dates dataframe foe ease of work. 
met_dates = met[['Object Date', 'Object Begin Date', 'Object End Date', 'Artist Begin Date', 'Artist End Date', 
                 'Artist Display Name' , 'Title']]

# View using head. 
met_dates.head(10)

Object Date  Object Begin Date  Object End Date Artist Begin Date  \
0        1853               1853             1853        1794         
1        1901               1901             1901        1785         
2     1909–27               1909             1927               NaN   
3     1909–27               1909             1927               NaN   
4     1909–27               1909             1927               NaN   
5     1909–27               1909             1927               NaN   
6     1909–27               1909             1927               NaN   
7     1909–27               1909             1927               NaN   
8     1909–27               1909             1927               NaN   
9        1912               1912             1912        1867         

  Artist End Date    Artist Display Name  \
0      1869        James Barton Longacre   
1      1844           Christian Gobrecht   
2             NaN                    NaN   
3             NaN                    NaN   
4             NaN                    NaN   
5             NaN                    NaN   
6             NaN                    NaN   
7             NaN                    NaN   
8             NaN                    NaN   
9      1917              Bela Lyon Pratt   

                                    Title  
0            One-dollar Liberty Head Coin  
1            Ten-dollar Liberty Head Coin  
2              Two-and-a-Half Dollar Coin  
3              Two-and-a-Half Dollar Coin  
4              Two-and-a-Half Dollar Coin  
5              Two-and-a-Half Dollar Coin  
6              Two-and-a-Half Dollar Coin  
7              Two-and-a-Half Dollar Coin  
8              Two-and-a-Half Dollar Coin  
9  Two-and-a-half-dollar Indian Head Coin

In [1539]:
# Print data types for the 3 date columns. 

# Loop through every column. 
for column in met_dates.columns:
    # Print Column name and data type. 
    print(f"Column: {column}, Data Type: {met_dates[column].dtype}")

Column: Object Date, Data Type: object
Column: Object Begin Date, Data Type: int64
Column: Object End Date, Data Type: int64
Column: Artist Begin Date, Data Type: object
Column: Artist End Date, Data Type: object
Column: Artist Display Name, Data Type: object
Column: Title, Data Type: object


Here we cann see that all of the dates here are not date time. lets conevrt all to strings and then to datetime data types. 

In [1541]:
# Convert all of the met_dates DataFrame to a string. 
# Use astype.
met_dates = met_dates.astype(str)

In [1542]:
# Verify DataTypes

# Loop through every column. 
for column in met_dates.columns:
    # Print Column name and data type. 
    print(f"Column: {column}, Data Type: {met_dates[column].dtype}")

Column: Object Date, Data Type: object
Column: Object Begin Date, Data Type: object
Column: Object End Date, Data Type: object
Column: Artist Begin Date, Data Type: object
Column: Artist End Date, Data Type: object
Column: Artist Display Name, Data Type: object
Column: Title, Data Type: object


Now we will convert the strings to date time data types. 

Additionally there appears to be multiple dates in the columns Artist Begin and  End Date. We will clean these columns by removing the secondary dates. 

In [1551]:
# Clean Columns

# Use str.strip to remove whitespace 
met_dates['Artist Begin Date'] = met_dates['Artist Begin Date'].str.strip()
# Use apply and lambda functions to take only the first 4 characters of each value in the Artist Begin column
met_dates['Artist Begin Date'] = met_dates['Artist Begin Date'].apply(lambda x: x[:4])


# Use str.strip to remove whitespace 
met_dates['Artist End Date'] = met_dates['Artist End Date'].str.strip()
# Use apply and lambda functions to take only the first 4 characters of each value in the Artist End column
met_dates['Artist End Date'] = met_dates['Artist End Date'].apply(lambda x: x[:4])


In [1553]:
met_dates

Object Date Object Begin Date Object End Date Artist Begin Date  \
0              1853              1853            1853              1794   
1              1901              1901            1901              1785   
2           1909–27              1909            1927               nan   
3           1909–27              1909            1927               nan   
4           1909–27              1909            1927               nan   
...             ...               ...             ...               ...   
484951         1944              1939            1949              1913   
484952         1944              1939            1949              1913   
484953  ca. 1925–33              1920            1938              1905   
484954         1839              1839            1839              1786   
484955         1887              1887            1887              1848   

       Artist End Date     Artist Display Name  \
0                 1869   James Barton Longacre   
1                 1844      Christian Gobrecht   
2                  nan                     nan   
3                  nan                     nan   
4                  nan                     nan   
...                ...                     ...   
484951            1985        Douglas Gorsline   
484952            1985        Douglas Gorsline   
484953            1969  Edmond Mario Granville   
484954            1889  Michel Eugène Chevreul   
484955            1948        Frank M. Gregory   

                                                    Title  
0                            One-dollar Liberty Head Coin  
1                            Ten-dollar Liberty Head Coin  
2                              Two-and-a-Half Dollar Coin  
3                              Two-and-a-Half Dollar Coin  
4                              Two-and-a-Half Dollar Coin  
...                                                   ...  
484951                                       Holiday Card  
484952                                     Brooklyn Local  
484953                                Yesterday and Today  
484954  De la loi du contraste simultané des couleurs,...  
484955  Madison Square - Fifth Avenue and Broadway at ...  

[484956 rows x 7 columns]

In [1555]:
# Import Packages. 
from datetime import datetime

In [1557]:
# Convert Object Begin to DateTime 
# use .apply(pd.to_datetime)
# errors = 'coerce' leaves null NaN
met_dates['Object Begin Date'] = met_dates['Object Begin Date'].apply(pd.to_datetime, format='%Y', errors='coerce')
                                                   

In [1558]:
# Convert Object End to DateTime
# use .appl(pd.to_datetime)
# errors = 'coerce' leaves null NaN

met_dates['Object End Date'] = met_dates['Object End Date'].apply(pd.to_datetime, format='%Y', errors='coerce')

In [1559]:
# Convert Artist End to DateTime
 # use .appl(pd.to_datetime)
# errors = 'coerce' leaves null NaN
met_dates['Artist End Date'] = met_dates['Artist End Date'].apply(pd.to_datetime, format='%Y', errors='coerce')

In [1560]:
# Convert Artist Begin to DateTime
 # use .appl(pd.to_datetime)
# errors = 'coerce' leaves null NaN
met_dates['Artist Begin Date'] = met_dates['Artist Begin Date'].apply(pd.to_datetime, format='%Y', errors='coerce')

In [1561]:
# Verify DataTypes

# Loop through every column. 
for column in met_dates.columns:
    # Print Column name and data type. 
    print(f"Column: {column}, Data Type: {met_dates[column].dtype}")

Column: Object Date, Data Type: object
Column: Object Begin Date, Data Type: datetime64[ns]
Column: Object End Date, Data Type: datetime64[ns]
Column: Artist Begin Date, Data Type: datetime64[ns]
Column: Artist End Date, Data Type: datetime64[ns]
Column: Artist Display Name, Data Type: object
Column: Title, Data Type: object


Here we can see that we have converted 'Object Begin Date' , 'Object End Date', 'Artist Begin Date' and 'Artist End Date' to datetime date types. This will be essential for our next exercise. 

# Generate Date Range

Here we will evaluate the career length of a particular artist and then display the date range of their career. We will then find the midpoint of that artists career and display all the pieces in the Met Museum data set created in that year. 

In [1564]:
# Filter out records where there is no artist begin or end date 
filt_met_dates = met_dates.dropna(subset=['Artist Begin Date'])
filt_met_dates = filt_met_dates.dropna(subset=['Artist Begin Date'])

In [1565]:
# Display df 
filt_met_dates.head(20)

Object Date Object Begin Date Object End Date Artist Begin Date  \
0          1853        1853-01-01      1853-01-01        1794-01-01   
1          1901        1901-01-01      1901-01-01        1785-01-01   
9          1912        1912-01-01      1912-01-01        1867-01-01   
10         1907        1907-01-01      1907-01-01        1785-01-01   
11         1876        1876-01-01      1876-01-01        1794-01-01   
12         1910        1910-01-01      1910-01-01        1867-01-01   
13         1907        1907-01-01      1907-01-01        1785-01-01   
32         1876        1876-01-01      1876-01-01        1861-01-01   
33      1847–50        1847-01-01      1850-01-01        1835-01-01   
34      1893–95        1893-01-01      1895-01-01        1859-01-01   
35     ca. 1887        1884-01-01      1887-01-01        1885-01-01   
36     ca. 1785        1782-01-01      1785-01-01        1765-01-01   
37     ca. 1785        1782-01-01      1785-01-01        1765-01-01   
39      1830–70        1830-01-01      1870-01-01        1818-01-01   
99    1792–1814        1792-01-01      1814-01-01        1736-01-01   
100   1792–1814        1792-01-01      1814-01-01        1736-01-01   
101     1803–28        1803-01-01      1828-01-01        1803-01-01   
102     1803–28        1803-01-01      1828-01-01        1803-01-01   
103    ca. 1810        1807-01-01      1810-01-01        1797-01-01   
104    ca. 1810        1807-01-01      1810-01-01        1797-01-01   

    Artist End Date                Artist Display Name  \
0        1869-01-01              James Barton Longacre   
1        1844-01-01                 Christian Gobrecht   
9        1917-01-01                    Bela Lyon Pratt   
10       1844-01-01                 Christian Gobrecht   
11       1869-01-01              James Barton Longacre   
12       1917-01-01                    Bela Lyon Pratt   
13       1844-01-01                 Christian Gobrecht   
32       1930-01-01           James Gillinder and Sons   
33       1853-01-01  Forestville Manufacturing Company   
34       1927-01-01     Paulding Farnham|Tiffany & Co.   
35       1932-01-01                   Conrad Henninger   
36       1795-01-01                     Pierre Stephan   
37       1795-01-01                     Pierre Stephan   
39       1888-01-01          New England Glass Company   
99       1815-01-01                        John Bailey   
100      1815-01-01                        John Bailey   
101      1828-01-01                        James Davis   
102      1828-01-01                        James Davis   
103      1820-01-01                      John Molineux   
104      1820-01-01                      John Molineux   

                                       Title  
0               One-dollar Liberty Head Coin  
1               Ten-dollar Liberty Head Coin  
9     Two-and-a-half-dollar Indian Head Coin  
10   Two-and-a-half-dollar Liberty Head Coin  
11           Twenty-dollar Liberty Head Coin  
12              Five-dollar Indian Head Coin  
13             Five-dollar Liberty Head Coin  
32                   Bust of Abraham Lincoln  
33                               Acorn Clock  
34                            The Adams Vase  
35                                Side Chair  
36           Figure of Admiral George Rodney  
37             Figure of Admiral Samuel Hood  
39                                 Ale Glass  
99                                   Andiron  
100                                  Andiron  
101                                  Andiron  
102                                  Andiron  
103                                  Andiron  
104                                  Andiron

Lets chose James Davis's career to evaluate. 

In [1590]:
# Now lets generate the year range from his career. 

# Use loc to find start date 
start_d = filt_met_dates.loc[filt_met_dates['Artist Display Name'] == 'James Davis', 'Artist Begin Date'].values[0]
print('Artist Start: ', start_d)

# Repeat to find end date
end_d = filt_met_dates.loc[filt_met_dates['Artist Display Name'] == 'James Davis', 'Artist End Date'].values[0]
print('Artist End: ', end_d)

# Generate year rand from career. 
year_range = pd.date_range(start=start_d, end=end_d, freq='Y')

# Print the result of the artists career year range 
print('James Davis Career Range in Years')
print(year_range)

Artist Start:  1803-01-01T00:00:00.000000000
Artist End:  1828-01-01T00:00:00.000000000
James Davis Career Range in Years
DatetimeIndex(['1803-12-31', '1804-12-31', '1805-12-31', '1806-12-31',
               '1807-12-31', '1808-12-31', '1809-12-31', '1810-12-31',
               '1811-12-31', '1812-12-31', '1813-12-31', '1814-12-31',
               '1815-12-31', '1816-12-31', '1817-12-31', '1818-12-31',
               '1819-12-31', '1820-12-31', '1821-12-31', '1822-12-31',
               '1823-12-31', '1824-12-31', '1825-12-31', '1826-12-31',
               '1827-12-31'],
              dtype='datetime64[ns]', freq='A-DEC')


In [1607]:
# Calculate the midpoint of the year range 
# First entry + (last entry -first entry) /2 
midpoint = year_range[0] + (year_range[-1] - year_range[0]) / 2

# Use strftime to format midpoint as year. 
midpoint_year = midpoint.strftime('%Y')

# Print 
print( 'The Midpoint of James Davis Career: ', midpoint)

The Midpoint of James Davis Career:  1815-12-31 00:00:00


In [1611]:
# Create Midpoint DataFrame
midpoint_df = filt_met_dates[filt_met_dates['Object Begin Date'] == midpoint_year]

# Display the pieces created in the midpoint of James Davis's Career
midpoint_df.head(15)

Object Date Object Begin Date Object End Date Artist Begin Date  \
274             1815–25        1815-01-01      1825-01-01        1783-01-01   
418                1815        1815-01-01      1815-01-01        1725-01-01   
1302        ca. 1810–30        1815-01-01      1835-01-01        1816-01-01   
1329            1815–20        1815-01-01      1820-01-01        1770-01-01   
1664            1815–20        1815-01-01      1820-01-01        1796-01-01   
1665            1815–20        1815-01-01      1820-01-01        1796-01-01   
1666            1815–20        1815-01-01      1820-01-01        1796-01-01   
1667            1815–20        1815-01-01      1820-01-01        1796-01-01   
1924            1815–20        1815-01-01      1820-01-01        1777-01-01   
2323            1815–35        1815-01-01      1835-01-01        1818-01-01   
2500            1815–35        1815-01-01      1835-01-01        1818-01-01   
2866            1815–41        1815-01-01      1841-01-01        1815-01-01   
2944            1815–20        1815-01-01      1820-01-01        1770-01-01   
3098  ca. 1818–ca. 1846        1815-01-01      1849-01-01        1818-01-01   
3100  ca. 1818–ca. 1846        1815-01-01      1849-01-01        1818-01-01   

     Artist End Date                                Artist Display Name  \
274       1864-01-01                                  Aaron Willard Jr.   
418       1792-01-01                              Jean Jacques Caffiéri   
1302      1851-01-01                               Henry Kellam Hancock   
1329      1826-01-01                                     Henry Connelly   
1664      1833-01-01                            John Finlay|Hugh Finlay   
1665      1833-01-01                            John Finlay|Hugh Finlay   
1666      1833-01-01                            John Finlay|Hugh Finlay   
1667      1833-01-01                            John Finlay|Hugh Finlay   
1924      1819-01-01                                              Dubuc   
2323      1888-01-01  New England Glass Company|South Boston Flint G...   
2500      1888-01-01  New England Glass Company|South Boston Flint G...   
2866      1815-01-01                                  Keene Glass Works   
2944      1854-01-01                                       Duncan Phyfe   
3098      1846-01-01                                  Enoch Wood & Sons   
3100      1846-01-01                                  Enoch Wood & Sons   

                           Title  
274                  Banjo Clock  
418                          Bus  
1302     Column Base and Capital  
1329                  Card Table  
1664                  Side Chair  
1665                  Side Chair  
1666                  Side Chair  
1667                  Side Chair  
1924                       Clock  
2323                  Sugar bowl  
2500               Cream pitcher  
2866                    Decanter  
2944  Cylinder Desk and Bookcase  
3098                        Dish  
3100                        Dish

In [1613]:
# Display the shape of midpoint_df
midpoint_df.shape

(541, 7)

As we can see there are 541 different pieces that were created in the midpoint of James Davis's Career